File: Leetcode/Python3/787. Cheapest Flights Within K Stops/solution.py
# https://leetcode.com/problems/cheapest-flights-within-k-stops/
# Ref: https://leetcode.com/problems/cheapest-flights-within-k-stops/solutions/115541/java-c-python-priority-queue-solution-tle-now/


from collections import defaultdict
import heapq


class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        graph = defaultdict(list)
        dist = [-float("inf") for _ in range(n + 1)]

        for f, t, p in flights:     # from, to, price
            graph[f].append((t, p))

        heap = [(0, k, src)]    # (price, distance, node)
        while heap:
            price, distance, node = heapq.heappop(heap)
            if node == dst:
                return price
            if distance < 0:
                continue
            # The larger distance, coming out of heapq first, so if it is lower than dist[node], skip it.
            if distance <= dist[node]:
                continue
            dist[node] = distance
            for v, w in graph[node]:
                heapq.heappush(heap, (price + w, distance - 1, v))

        return -1

File: Leetcode/C++/787. Cheapest Flights Within K Stops/solution.cpp
// https://leetcode.com/problems/cheapest-flights-within-k-stops/

#include <vector>
#include <climits>
#include <queue>

using namespace std;

class Solution
{
public:
    int findCheapestPrice(int n, vector<vector<int>> &flights, int src, int dst, int k)
    {
        vector<int> distances(n, INT_MAX);
        vector<vector<pair<int, int>>> graph(n);
        for (vector<int> vec : flights)
        {
            int from = vec[0], to = vec[1], price = vec[2];
            graph[from].push_back({price, to});
        }

        priority_queue<tuple<int, int, int>> pq;
        pq.push({0, src, 0}); // distance, city, stops
        int newDist, nextCity;
        while (!pq.empty())
        {
            int dist = get<0>(pq.top()), city = get<1>(pq.top()), stops = get<2>(pq.top());
            pq.pop();
            if (distances[city] > dist)
            {
                distances[city] = dist;
                for (auto p : graph[city])
                {
                    newDist = dist + p.first;
                    nextCity = p.second;
                    // If distances[nextCity] is bigger than newDist and nextcity <= k stops push to priority_queue for search
                    if (distances[nextCity] > newDist && stops <= k)
                    {
                        pq.push({newDist, nextCity, stops + 1});
                    }
                }
            }
        }

        return distances[dst] == INT_MAX ? -1 : distances[dst];
    }
};